test/top-story-endpoint

fperedo opened the pull request:
This PR introduces comprehensive test coverage for the Top Stories API endpoints, adds functionality to the APIClient utility class, and extends the Assertions module with advanced validation methods. The following changes aim to improve test reliability, ensure schema validation, and enhance debugging capabilities.

**New Tests for Top Stories API**
File: tests/test_top_stories.py
Purpose: Provide end-to-end testing for GET, POST, PATCH, and specific GET requests related to the Top Stories API.
Highlights:
Parametrized test cases for POST and PATCH requests using test data from JSON files.
Validation of API responses against predefined schemas (GET_TOP_STORIES_SCHEMA, TOP_STORY_SCHEMA).
Assertions for status codes, nested data structures, and expected response values.

**Enhancements to API Client**
File: utils/api_client.py
Change: Added a patch method to support sending PATCH requests.
Purpose: Extend API client capabilities for testing and interacting with APIs requiring partial updates.
Method Details:
Accepts data and files parameters for flexibility in request payloads.
Handles headers automatically for consistent testing.

**Extended Assertions Module**
File: utils/assertions.py
Change: Introduced assert_nested_contains method.
Purpose: Enable recursive validation of nested dictionaries and lists, ensuring comprehensive data checks in deeply structured responses.
Highlights:
Detailed error messages with path-specific reporting for easier debugging.
Supports mixed data types (e.g., dicts within lists).

**Testing Instructions**
Ensure that all required dependencies (pytest, dotenv, etc.) are installed.
Run the test suite using:

```
pytest .\tests\test_top_stories.py
```
Confirm that all tests pass successfully.

![image](https://github.com/user-attachments/assets/ef6d557e-ee09-42d0-811f-a2a7f78938c6)

daviddflix commented:
Hey @fperedo , I noticed 6 failed tests. Could you take a look at this? Let me know if you need more details from my side!

![image](https://github.com/user-attachments/assets/2a2d93a3-bead-467a-8cda-041129b796de)

fperedo commented:
- Added a new test for auth0id validation on POST /users endpoint
- Fixed POST /top-story tests by modifying the articles id
@daviddflix Please check again! Thanks

![image](https://github.com/user-attachments/assets/e7d3bfc4-f5fa-4bc9-aa43-cc1013002d6a)
